core/arch/arm/tee/entry_fast.c

```diff
@@ -102,7 +102,19 @@ static void tee_entry_exchange_capabilities(struct thread_smc_args *args)
 	}
 
 	args->a0 = OPTEE_SMC_RETURN_OK;
+
+#ifdef CFG_SMALL_PAGE_USER_TA
 	args->a1 = OPTEE_SMC_SEC_CAP_HAVE_RESERVED_SHM;
+	if (core_mmu_is_nsec_ddr_configured()) {
+		IMSG("NS DDR configured. Enabling dynamic SHM");
+		args->a1 |= OPTEE_SMC_SEC_CAP_UNREGISTERED_SHM |
+			OPTEE_SMC_SEC_CAP_REGISTER_SHM;
+	} else {
+		EMSG("No NS DDR configured for the platform. Disabling dynamic SHM");
+	}
+#else
```

Comment:
We wouldn't need this `#else` if the `#ifdef CFG_SMALL_PAGE_USER_TA` above was moved down one line.